python/ray/autoscaler/_private/autoscaler.py

```diff
@@ -478,50 +504,65 @@ def _get_nodes_allowed_to_terminate(
             worker's node_type.
 
-        Also updates and returns the dictionary of node type counts.
+        Returns KeepOrTerminate.terminate when either the node type is not
+        among available node type, or both the following hold:
+        (a) The worker's node_type is present among the keys of the current
+            config's available_node_types dict.
+        (b) Keeping the node would violate the max_workers constraint for that
+            worker's node_type.
+
+        Return KeepOrTerminate.decide_later otherwise.
+
 
         Args:
             node_type_counts(Dict[NodeType, int]): The non_terminated node
                 types counted so far.
         Returns:
-            bool: True if the node should be kept. False otherwise.
-            Dict[NodeType, int]: Updated node type counts
+            KeepOrTerminate: keep if the node should be kept, terminate if the
+            node should be terminated, decide_later if we are allowed
+            to terminate it, but do not have to.
+            Optional[str]: reason for termination.
         """
-        new_node_type_counts = copy.deepcopy(node_type_counts)
         tags = self.provider.node_tags(node_id)
         if TAG_RAY_USER_NODE_TYPE in tags:
             node_type = tags[TAG_RAY_USER_NODE_TYPE]
-            if node_type not in self.available_node_types:
-                # The node type has been deleted from the cluster config.
-                # Don't keep the node.
-                return False, new_node_type_counts
-            new_node_type_counts[node_type] += 1
-            min_workers = self.available_node_types[node_type].get(
+            head_node_type = self.config["head_node_type"]
+
+            min_workers = self.available_node_types.get(node_type, {}).get(
                 "min_workers", 0)
-            max_workers = self.available_node_types[node_type].get(
+            max_workers = self.available_node_types.get(node_type, {}).get(
                 "max_workers", 0)
-            if new_node_type_counts[node_type] <= min(min_workers,
-                                                      max_workers):
-                return True, new_node_type_counts
-
-        return False, new_node_type_counts
+            if node_type not in self.available_node_types:
+                # The node type has been deleted from the cluster config.
+                # Allow terminating it if needed.
+                available_node_types = list(self.available_node_types.keys())
+                return (KeepOrTerminate.terminate,
+                        f"not in available_node_types: {available_node_types}")
+            new_count = node_type_counts[node_type] + 1
+            if (new_count <= min(min_workers, max_workers)
+                    or node_type == head_node_type):
+                return KeepOrTerminate.keep, None
+            if new_count > max_workers:
+                return KeepOrTerminate.terminate, "max_workers_per_type"
```

Comment:
oh, realized there's this annoying issue:
because we go through the nodes starting from the most recently used,
this prioritizes killing most recently used nodes first, which is the opposite of what's desired. 

If there's a straightforward way to fix the prioritization for terminating above per-type max workers, that would be great.
If it's too much of a mess to fix, that's fine -- perfect behavior under reduction of max_workers is not the biggest concern right now.

Reply:
I do not see the issue.

We go through the nodes, starting from most recently used, keeping the nodes until we accumulate max_workers per type, then start killing nodes of the given type.
This seems to prioritize _keeping_ most recently used nodes first.

Reply:
A great discussion. Yes, I think we keep the most recently used and then we start removing from the least recently used but not in the optimal order because ideally in a case like this:
[12 AM, 11 AM, 10 AM, 9AM] where you want to start removing from 10AM and earlier, you would first remove 10 AM and then 9 AM. Ideally it would be 9AM and then 10 AM. So I think both of you are correct.

Reply:
Sasha is correct -- say you have nodes of a given type last used at 
[12, 11, 10, 9]
min_workers = 1, max_workers = 2 
12 is kept, 11 is decided later, and the rest (9, 10) are terminated later.

The loop's decision to remove 10 comes before the decision to remove 9, but that doesn't matter.